// @mr-hope/vuepress-plugin-feed/lib/node/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCategory = exports.getAuthor = exports.FEED_GENERATOR = exports.encodeXML = exports.encodeCDATA = exports.getImageMineType = exports.resolveUrl = exports.isAbsoluteUrl = exports.isUrl = exports.resolveHTML = exports.compareDate = exports.success = exports.error = exports.info = exports.warn = exports.wait = void 0;
const chalk_1 = require("chalk");
const wait = (message) => console.log((0, chalk_1.blue)("Feed:"), chalk_1.black.bgYellowBright("wait"), message);
exports.wait = wait;
const warn = (message) => console.log((0, chalk_1.blue)("Feed:"), chalk_1.black.bgYellow("warn"), message);
exports.warn = warn;
const info = (message) => console.log((0, chalk_1.blue)("Feed:"), chalk_1.black.bgBlue("Info"), message);
exports.info = info;
const error = (message) => console.log((0, chalk_1.blue)("Feed:"), chalk_1.black.bgRed("Error"), message);
exports.error = error;
const success = (message) => console.log((0, chalk_1.blue)("Feed:"), chalk_1.black.bgGreen("Success"), message);
exports.success = success;
const compareDate = (dateA, dateB) => {
    if (!dateA || !(dateA instanceof Date))
        return 1;
    if (!dateB || !(dateB instanceof Date))
        return -1;
    return dateB.getTime() - dateA.getTime();
};
exports.compareDate = compareDate;
const resolveHTML = (html, customElements = []) => html
    // remove html class
    .replace(/ class=".*?"/gu, "")
    // remove v-pre
    .replace(/ v-pre/gu, "")
    // remove anchor
    .replace(/<a href="#.*?">.*?<\/a>/gu, "")
    // remove html comment
    .replace(/(<!--.*?--!?>)|(<!--[\S\s]+?--!?>)|(<!--[\S\s]*?$)/gu, "")
    // remove ExternalLinkIcon
    .replace(/<ExternalLinkIcon ?\/>/gu, "")
    // resolve RouterLink
    .replace(/<RouterLink to="(.*?)">(.*?)<\/RouterLink>/gu, '<a href="$1">$2</a>')
    // remove self-closed tags
    .replace(/<(?:a|div|span)[^>]*?\/>/gu, "")
    // remove other related tags
    .replace(new RegExp(`<(${customElements.join("|")})[^>]*?(?:>*?<\\/\\1>|\\/>)`, "gu"), "<i>Content not supported</i>")
    // remove tex
    .replace(/<math[\s\S]*?\/math>/gu, "<i>Content not supported</i>");
exports.resolveHTML = resolveHTML;
/**
 * check if string is a valid url
 */
const isUrl = (test) => {
    if (typeof test !== "string" || test === "")
        return false;
    // url Math
    const result = /^(?:\w+:)?\/\/(\S+)$/u.exec(test);
    if (!result)
        return false;
    const address = result[1];
    if (!address)
        return false;
    return (
    // address with localhost
    /^localhost[:?\d]*(?:[^:?\d]\S*)?$/u.test(address) ||
        // address without localhost
        /^[^\s.]+\.\S{2,}$/u.test(address));
};
exports.isUrl = isUrl;
const isAbsoluteUrl = (test) => test.startsWith("/");
exports.isAbsoluteUrl = isAbsoluteUrl;
const resolveUrl = (hostname, base = "", path = "") => `${hostname}${base}${
// make sure path does not start with '/'
path.replace(/^\//u, "")}`;
exports.resolveUrl = resolveUrl;
const getImageMineType = (ext = "") => `image/${ext === "jpg"
    ? "jpeg"
    : ext === "svg"
        ? "svg+xml"
        : ext === "jpeg" ||
            ext === "png" ||
            ext === "bmp" ||
            ext === "gif" ||
            ext === "webp"
            ? ext
            : ""}`;
exports.getImageMineType = getImageMineType;
/**
 * @see https://stackoverflow.com/questions/223652/is-there-a-way-to-escape-a-cdata-end-token-in-xml
 */
const encodeCDATA = (content) => content.replace(/]]>/g, "]]]]><![CDATA[>");
exports.encodeCDATA = encodeCDATA;
const encodeXML = (content) => content
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;")
    .replace(/'/g, "&apos;");
exports.encodeXML = encodeXML;
exports.FEED_GENERATOR = "@mr-hope/vuepress-plugin-feed";
const getAuthor = (author, canDisable = false) => {
    if (author) {
        if (Array.isArray(author)) {
            return author.map((item) => typeof item === "string" ? { name: item } : item);
        }
        if (typeof author === "string")
            return [{ name: author }];
        if (typeof author === "object" && author.name)
            return [author];
        console.error(`Expect 'author' to be \`AuthorInfo[] | AuthorInfo | string[] | string ${canDisable ? "" : "| false"} | undefined\`, but got`, author);
        return [];
    }
    return [];
};
exports.getAuthor = getAuthor;
const getCategory = (category) => {
    if (category) {
        if (Array.isArray(category))
            return category;
        if (typeof category === "string")
            return [category];
        console.error(`Expect 'category' to be \`string[] | string | undefined\`, but got`, category);
    }
    return [];
};
exports.getCategory = getCategory;
