// @mr-hope/vuepress-plugin-components/lib/node/index.js
"use strict";
const vuepress_shared_1 = require("@mr-hope/vuepress-shared");
const path_1 = require("path");
const locales_1 = require("./locales");
const noopModule = "@mr-hope/vuepress-shared/lib/esm/noopModule";
const componentPlugin = (options, context) => {
    const plugins = [];
    if (options.pageinfo)
        plugins.push(["@mr-hope/git", true], ["@mr-hope/reading-time", { wordPerminute: options.wordPerminute }]);
    return {
        name: "@mr-hope/vuepress-plugin-components",
        alias: {
            "@BackToTop": options.backToTop
                ? (0, path_1.resolve)(__dirname, "../client/BackToTop.vue")
                : noopModule,
            "@BreadCrumb": options.breadcrumb
                ? (0, path_1.resolve)(__dirname, "../client/BreadCrumb.vue")
                : noopModule,
            "@Badge": options.badge
                ? (0, path_1.resolve)(__dirname, "../client/Badge.vue")
                : noopModule,
            "@PageInfo": options.pageinfo
                ? (0, path_1.resolve)(__dirname, "../client/PageInfo.vue")
                : noopModule,
            "@Pagination": options.pagination
                ? (0, path_1.resolve)(__dirname, "../client/Pagination.vue")
                : noopModule,
            "@ScreenFull": options.screenFull
                ? (0, path_1.resolve)(__dirname, "../client/ScreenFull.vue")
                : noopModule,
        },
        define: () => ({
            BACK_TO_TOP_THRESHOLD: options.backToTopThreshold || 300,
            COMPONENT_LOCALES: (0, vuepress_shared_1.getLocales)(context, locales_1.componentLocales, options.locales),
            PAGE_INFO_LOCALES: (0, vuepress_shared_1.getLocales)(context, locales_1.pageInfoLocales, options.pageInfoLocales),
            PAGINATION_LOCALES: (0, vuepress_shared_1.getLocales)(context, locales_1.paginationLocales, options.paginationLocales),
        }),
        enhanceAppFiles: (0, path_1.resolve)(__dirname, "../client/enhanceAppFile.js"),
        globalUIComponents: options.backToTop ? "BackToTop" : [],
        plugins,
    };
};
module.exports = componentPlugin;
